lib/interpolators.js

```diff
@@ -0,0 +1,173 @@
+// Return the value of an opportunity density grid cell.
+// Range check and return 0 for cells outside the grid.
+// TODO This should really be in the Grid class, and grids should have a default value for empty cells.
+function getGridValue (grid, gx, gy) {
+  if (gx < 0 || gx >= grid.width || gy < 0 || gy >= grid.height) return 0
+  // Convert grid x and y cell coordinates to a 1D offset into the grid.
+  // This is a good place to apply log or sqrt transformations if needed.
+  return grid.data[gy * grid.width + gx]
+}
+
+// An instance of this class represents a curve fit to four uniformly spaced data points.
+// Given four adjacent values a, b, c, d, it provides interpolated values between b and c using a and d to 
+// determine the slope going into and out of the b-c interval.
+// This is only used internally by the bicubic interpolator so it's not exported.
+class CubicHermiteInterpolator {
+    
+  constructor (a, b, c, d) {
+    this.c3 = -a / 2.0 + (3.0 * b) / 2.0 - (3.0 * c) / 2.0 + d / 2.0
+    this.c2 = a - (5.0 * b) / 2.0 + 2.0 * c - d / 2.0
+    this.c1 = -a / 2.0 + c / 2.0;
+    this.c0 = b;
+  }
+
+  // Given a value in [0, 1] expressing the position between b and c, return the interpolated value.
+  interpolate (fraction) {
+    return this.c3 * fraction ** 3 + this.c2 * fraction ** 2 + this.c1 * fraction + this.c0;
+  }
+     
+}
+
+// A pre-calculated bicubic interpolation patch.
+// For a 4x4 grid of samples, this allows us to calculate interpolated values between the central four samples.
+// By pre-fitting the curves in one dimension (y) and proceeding with the interpolation row by row, 
+// we re-use most of the computation from one output pixel to the next.
+export class BicubicInterpolator {
+
+  static tileOffset = 0.5
+
+  // Supply an unrolled row-major grid of 16 values (a 4x4 grid).
+  // The resulting object can be used to interpolate between the inner four cells.
+  // Maybe we should be initializing this with a typed 2D array instead of this mess of individual variables.
+  constructor (p00, p01, p02, p03, p10, p11, p12, p13, p20, p21, p22, p23, p30, p31, p32, p33) {
```

Comment:
For any function with more than two parameters I recommend using options objects.

```js
constructor (grid) {
  this.columnInterpolator0 = new CubicHermiteInterpolator(grid.p00, grid.p01, grid.p02, grid.p03)
  ...
}
```